### PR TITLE
lumper: add serial console file option

### PR DESCRIPTION
**Pull request by alexandrebrg**
As discussed today, here is the feature which redirect stdout from guest to a file. 

@sameo

### CODE REVIEW

File: lumper/src/vmm/src/lib.rs [hunk 1]
```diff
@@ -160,6 +167,26 @@ impl VMM {
...
+            // We create the file if it does not exist, else we open
+            let file = OpenOptions::new().
+                append(true).
+                create(true).
+                open(console_path).
+                map_err(ConsoleError)?;
+
+            self.serial = Arc::new(Mutex::new(
+                LumperSerial::new(Box::new(file)).map_err(Error::SerialCreation)?,
+            ));
```
Comment (alexandrebrg) [inline lumper/src/vmm/src/lib.rs]: About these three lines, I feel like it's not something great. I'd rather update the inner value of the Mutex than this.

I found an alternative to this, but I think it's even worse than the current solution:

```rust
            let mut serial = self.serial.lock().unwrap();
            // this works
            mem::replace(&mut *serial, LumperSerial::new(Box::new(file)).map_err(Error::SerialCreation)?);
```


To finish, I tried a third solution, where I was just updating the value by assignation, but I couldn't make it work.

```rust
            let mut serial = self.serial.lock().unwrap();
            // this does not work
            *serial = LumperSerial::new(Box::new(file)).map_err(Error::SerialCreation)?;
```

File: lumper/src/vmm/src/devices/serial.rs [hunk 2]
```diff
@@ -1,10 +1,10 @@
 // SPDX-License-Identifier: Apache-2.0
 
-use std::io::{stdout, Error, Result, Stdout};
+use std::io::{Error, Result, Write};
 use std::ops::Deref;
 
-use vm_superio::serial::NoEvents;
 use vm_superio::{Serial, Trigger};
+use vm_superio::serial::NoEvents;
```
Comment (sameo) [inline lumper/src/vmm/src/devices/serial.rs:6]: This change is not needed.

File: lumper/src/vmm/src/devices/serial.rs [hunk 3]
```diff
@@ -41,16 +41,15 @@ pub(crate) struct LumperSerial {
...
     // serial is the actual serial device.
-    pub serial: Serial<EventFdTrigger, NoEvents, Stdout>,
+    pub serial: Serial<EventFdTrigger, NoEvents, Box<dyn Write + Send>>,
 }
 
 impl LumperSerial {
-    pub fn new() -> Result<Self> {
+    pub fn new(bind: Box<dyn Write + Send>) -> Result<Self> {
         let eventfd = EventFdTrigger::new(libc::EFD_NONBLOCK).unwrap();
-
```
Comment (sameo) [inline lumper/src/vmm/src/devices/serial.rs:50]: Not needed as well.

File: lumper/src/vmm/src/devices/serial.rs [hunk 4]
```diff
@@ -41,16 +41,15 @@ pub(crate) struct LumperSerial {
     eventfd: EventFdTrigger,
 
     // serial is the actual serial device.
-    pub serial: Serial<EventFdTrigger, NoEvents, Stdout>,
+    pub serial: Serial<EventFdTrigger, NoEvents, Box<dyn Write + Send>>,
 }
 
 impl LumperSerial {
-    pub fn new() -> Result<Self> {
+    pub fn new(bind: Box<dyn Write + Send>) -> Result<Self> {
```
Comment (sameo) [inline lumper/src/vmm/src/devices/serial.rs]: I'd call it `output` instead of `bind`

File: lumper/src/vmm/src/lib.rs [hunk 5]
```diff
@@ -8,24 +8,29 @@ extern crate linux_loader;
...
+use std::fs::{OpenOptions};
+use std::io::stdout;
 use std::os::unix::io::AsRawFd;
 use std::os::unix::prelude::RawFd;
 use std::sync::{Arc, Mutex};
 use std::thread;
-use std::{io, path::PathBuf};
 
-use kvm_bindings::{kvm_userspace_memory_region, KVM_MAX_CPUID_ENTRIES};
+use kvm_bindings::{KVM_MAX_CPUID_ENTRIES, kvm_userspace_memory_region};
```
Comment (sameo) [inline lumper/src/vmm/src/lib.rs]: Unneeded change.

File: lumper/src/vmm/src/lib.rs [hunk 6]
```diff
@@ -8,24 +8,29 @@ extern crate linux_loader;
...
-mod cpu;
 use cpu::{cpuid, mptable, Vcpu};
-mod devices;
 use devices::serial::LumperSerial;
+use epoll_context::{EPOLL_EVENTS_LEN, EpollContext};
+
+use crate::Error::ConsoleError;
+
+mod cpu;
+mod devices;
```
Comment (sameo) [inline lumper/src/vmm/src/lib.rs]: This header reorganization should not be part of this patch.

File: lumper/src/vmm/src/lib.rs (hunk 1 above)
Comment (sameo) [reply lumper/src/vmm/src/lib.rs]: > ```rust
>             let mut serial = self.serial.lock().unwrap();
>             // this does not work
>             *serial = LumperSerial::new(Box::new(file)).map_err(Error::SerialCreation)?;
> ```

This is the idiomatic way of modifying the inner value of a mutex, and it does work. What's not working for you?

File: lumper/src/vmm/src/lib.rs [hunk 7]
```diff
@@ -160,6 +167,26 @@ impl VMM {
...
+        console_path: Option<String>
+    ) -> Result<()> {
+        if let Some(console_path) = console_path {
+            // We create the file if it does not exist, else we open
+            let file = OpenOptions::new().
+                append(true).
+                create(true).
+                open(console_path).
+                map_err(ConsoleError)?;
+
```
Comment (sameo) [inline lumper/src/vmm/src/lib.rs]: You could simply do a

```rust
let console = File::create(&console_path).map_error(ConsoleError);
```

This will truncate the file if it already exists, which is the behaviour we're looking for.

**Review by sameo — COMMENTED** [review]
Thanks for the PR @AlexandreBrg.
I have a few rather minor comments.

Also, please add a SOB to your commit message, together with at least a short sentence describing what you're doing (and the new CLI option).

File: lumper/src/vmm/src/lib.rs (hunk 7 above)
Comment (alexandrebrg) [reply lumper/src/vmm/src/lib.rs]: In fact, I thought the behavior we wanted was to make an update to the console file. I did the update with the new behavior.

File: lumper/src/vmm/src/lib.rs (hunk 1 above)
Comment (alexandrebrg) [reply lumper/src/vmm/src/lib.rs]: I was pretty sure it did not work yesterday when I tried. Whatever, now it's working as intended, I did the update.

**Comment by alexandrebrg** [comment]
> Thanks for the PR @AlexandreBrg. I have a few rather minor comments.
> 
> Also, please add a SOB to your commit message, together with at least a short sentence describing what you're doing (and the new CLI option).

I updated the code, also for most import changes, I guess it's my IDE which was messing with them (I'll take a look into it). I also did the update for the commit, about that, why is the SOB required? 

Thanks for the review

**Review by sameo — APPROVED** [review]

**Comment by sameo** [comment]
> Why is the SOB required?

It is required to track copyright and it is an implicit [developer certificate](https://developercertificate.org/). Typically projects assigns the SOB to the certificate by documenting it, but here it's implicit.
